File: examples/typescript/part-4/12-hallucination-detection.ts
// Illustrative examples for notes/part-4-output-security/12-hallucination-detection.md
// Not for production use. Licensed under MIT (see LICENSE-CODE).

enum ClaimType {
  Fact = "fact",
  Number = "number",
  Citation = "citation",
  ToolResult = "tool_result",
  Policy = "policy",
  Opinion = "opinion",
}

enum ClaimStatus {
  Supported = "supported",
  Unsupported = "unsupported",
  Contradicted = "contradicted",
  Stale = "stale",
  NotCheckable = "not_checkable",
}

interface Claim {
  id: string;
  text: string;
  type: ClaimType;
  sourceIds: string[];
  createdAt: Date;
}

interface ClaimResult {
  claimId: string;
  status: ClaimStatus;
  evidenceId: string;
  reason: string;
  confidence: number;
}

interface Evidence {
  id: string;
  text: string;
  freshAt: string;
  acl: string[];
}

interface EvidenceStore {
  find(claim: Claim): Evidence[];
}

class ClaimVerifier {
  store: EvidenceStore;

  constructor(store: EvidenceStore) {
    this.store = store;
  }

  verify(claim: Claim): ClaimResult {
    if (claim.type === ClaimType.Opinion) {
      return {
        claimId: claim.id,
        status: ClaimStatus.NotCheckable,
        evidenceId: "",
        reason: "opinion",
        confidence: 0,
      };
    }

    const evidence = this.store.find(claim);
    if (evidence.length === 0) {
      return {
        claimId: claim.id,
        status: ClaimStatus.Unsupported,
        evidenceId: "",
        reason: "no evidence found",
        confidence: 0,
      };
    }

    for (const item of evidence) {
      if (supports(claim.text, item.text)) {
        return {
          claimId: claim.id,
          status: ClaimStatus.Supported,
          evidenceId: item.id,
          reason: "matched evidence",
          confidence: 0.8,
        };
      }
    }

    return {
      claimId: claim.id,
      status: ClaimStatus.Unsupported,
      evidenceId: "",
      reason: "evidence found but no support",
      confidence: 0.4,
    };
  }
}

function supports(claimText: string, evidenceText: string): boolean {
  // Упрощённый пример. В реальной системе здесь может быть:
  // - exact match для чисел/дат/id;
  // - semantic similarity;
  // - отдельный verifier model;
  // - проверка по БД/API.
  return evidenceText.toLowerCase().includes(claimText.toLowerCase());
}

enum PublishDecision {
  Allow = "allow",
  Rewrite = "rewrite",
  Block = "block",
  Review = "review",
}

class HallucinationPolicy {
  blockContradictions: boolean;
  reviewHighImpact: boolean;

  constructor(blockContradictions = true, reviewHighImpact = true) {
    this.blockContradictions = blockContradictions;
    this.reviewHighImpact = reviewHighImpact;
  }

  decide(results: ClaimResult[], highImpact: boolean): PublishDecision {
    for (const result of results) {
      if (result.status === ClaimStatus.Contradicted && this.blockContradictions) {
        throw new Error(`contradicted claim: ${result.claimId}`);
      }
    }

    if (highImpact && this.reviewHighImpact) {
      for (const result of results) {
        if (
          result.status === ClaimStatus.Unsupported ||
          result.status === ClaimStatus.NotCheckable
        ) {
          throw new Error(
            `high-impact answer contains unverified claim: ${result.claimId}`,
          );
        }
      }
    }

    for (const result of results) {
      if (
        result.status === ClaimStatus.Unsupported ||
        result.status === ClaimStatus.Stale
      ) {
        return PublishDecision.Rewrite;
      }
    }

    return PublishDecision.Allow;
  }
}

function rewriteWithUncertainty(answer: string, results: ClaimResult[]): string {
  const needsUncertainty = results.some(
    (result) =>
      result.status === ClaimStatus.Unsupported ||
      result.status === ClaimStatus.Stale ||
      result.status === ClaimStatus.NotCheckable,
  );

  if (!needsUncertainty) {
    return answer;
  }

  const prefix = "По доступным источникам это не подтверждено полностью. ";
  if (answer.startsWith(prefix)) {
    return answer;
  }
  return prefix + answer;
}

File: examples/python/part-4/12-hallucination-detection.py
# Illustrative examples for notes/part-4-output-security/12-hallucination-detection.md
# Not for production use. Licensed under MIT (see LICENSE-CODE).

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import List, Protocol


class ClaimType(str, Enum):
    FACT = "fact"
    NUMBER = "number"
    CITATION = "citation"
    TOOL_RESULT = "tool_result"
    POLICY = "policy"
    OPINION = "opinion"


class ClaimStatus(str, Enum):
    SUPPORTED = "supported"
    UNSUPPORTED = "unsupported"
    CONTRADICTED = "contradicted"
    STALE = "stale"
    NOT_CHECKABLE = "not_checkable"


@dataclass
class Claim:
    id: str
    text: str
    type: ClaimType
    source_ids: List[str] = field(default_factory=list)
    created_at: datetime = field(default_factory=datetime.utcnow)


@dataclass
class ClaimResult:
    claim_id: str
    status: ClaimStatus
    evidence_id: str = ""
    reason: str = ""
    confidence: float = 0.0


@dataclass
class Evidence:
    id: str
    text: str
    fresh_at: str = ""
    acl: List[str] = field(default_factory=list)


class EvidenceStore(Protocol):
    def find(self, claim: Claim) -> List[Evidence]: ...


@dataclass
class ClaimVerifier:
    store: EvidenceStore

    def verify(self, claim: Claim) -> ClaimResult:
        if claim.type == ClaimType.OPINION:
            return ClaimResult(
                claim_id=claim.id,
                status=ClaimStatus.NOT_CHECKABLE,
                reason="opinion",
            )

        evidence = self.store.find(claim)
        if not evidence:
            return ClaimResult(
                claim_id=claim.id,
                status=ClaimStatus.UNSUPPORTED,
                reason="no evidence found",
            )

        for item in evidence:
            if supports(claim.text, item.text):
                return ClaimResult(
                    claim_id=claim.id,
                    status=ClaimStatus.SUPPORTED,
                    evidence_id=item.id,
                    reason="matched evidence",
                    confidence=0.8,
                )

        return ClaimResult(
            claim_id=claim.id,
            status=ClaimStatus.UNSUPPORTED,
            reason="evidence found but no support",
            confidence=0.4,
        )


def supports(claim_text: str, evidence_text: str) -> bool:
    # Упрощённый пример. В реальной системе здесь может быть:
    # - exact match для чисел/дат/id;
    # - semantic similarity;
    # - отдельный verifier model;
    # - проверка по БД/API.
    return claim_text.lower() in evidence_text.lower()


class PublishDecision(str, Enum):
    ALLOW = "allow"
    REWRITE = "rewrite"
    BLOCK = "block"
    REVIEW = "review"


@dataclass
class HallucinationPolicy:
    block_contradictions: bool = True
    review_high_impact: bool = True

    def decide(
        self, results: List[ClaimResult], high_impact: bool
    ) -> PublishDecision:
        for result in results:
            if result.status == ClaimStatus.CONTRADICTED and self.block_contradictions:
                raise ValueError(f"contradicted claim: {result.claim_id}")

        if high_impact and self.review_high_impact:
            for result in results:
                if result.status in (
                    ClaimStatus.UNSUPPORTED,
                    ClaimStatus.NOT_CHECKABLE,
                ):
                    raise ValueError(
                        f"high-impact answer contains unverified claim: {result.claim_id}"
                    )

        for result in results:
            if result.status in (ClaimStatus.UNSUPPORTED, ClaimStatus.STALE):
                return PublishDecision.REWRITE

        return PublishDecision.ALLOW


def rewrite_with_uncertainty(answer: str, results: List[ClaimResult]) -> str:
    needs_uncertainty = any(
        result.status
        in (
            ClaimStatus.UNSUPPORTED,
            ClaimStatus.STALE,
            ClaimStatus.NOT_CHECKABLE,
        )
        for result in results
    )

    if not needs_uncertainty:
        return answer

    prefix = "По доступным источникам это не подтверждено полностью. "
    if answer.startswith(prefix):
        return answer
    return prefix + answer
